subsys/net/lib/dhcp_server/dhcpv4_server.c

```diff
@@ -0,0 +1,1423 @@
+				struct dhcpv4_server_ctx *ctx,
+				struct dhcpv4_parameter_request_list *params)
+{
+	for (uint8_t i = 0; i < params->count; i++) {
+		switch (params->list[i]) {
+		case DHCPV4_OPTIONS_SUBNET_MASK:
+			buf = dhcpv4_encode_subnet_mask_option(
+				buf, buflen, &ctx->netmask);
+			if (buf == NULL) {
+				goto out;
+			}
+			break;
+
+		/* Others - just ignore. */
+		default:
+			break;
+		}
+	}
+
+out:
+	return buf;
+}
+
+static int dhcpv4_send(struct dhcpv4_server_ctx *ctx, enum net_dhcpv4_msg_type type,
+		       uint8_t *reply, size_t len, struct dhcp_msg *msg,
+		       struct in_addr *yiaddr)
+{
+	struct sockaddr_in dst_addr = {
+		.sin_family = AF_INET,
+		.sin_port = htons(DHCPV4_CLIENT_PORT),
+	};
+	struct in_addr giaddr; /* Relay agent address */
+	struct in_addr ciaddr; /* Client address */
+	int ret;
+
+	memcpy(&giaddr, msg->giaddr, sizeof(giaddr));
+	memcpy(&ciaddr, msg->ciaddr, sizeof(ciaddr));
+
+	/* Select destination address as described in ch. 4.1. */
+	if (!net_ipv4_is_addr_unspecified(&giaddr)) {
+		/* If the 'giaddr' field in a DHCP message from a client is
+		 * non-zero, the server sends any return messages to the
+		 * 'DHCP server' port on the BOOTP relay agent whose address
+		 * appears in 'giaddr'.
+		 */
+		dst_addr.sin_addr = giaddr;
+		dst_addr.sin_port = htons(DHCPV4_SERVER_PORT);
+	} else if (type == NET_DHCPV4_MSG_TYPE_NAK) {
+		/* In all cases, when 'giaddr' is zero, the server broadcasts
+		 * any DHCPNAK messages to 0xffffffff.
+		 */
+		dst_addr.sin_addr = *net_ipv4_broadcast_address();
+	} else if (!net_ipv4_is_addr_unspecified(&ciaddr)) {
+		/* If the 'giaddr' field is zero and the 'ciaddr' field is
+		 * nonzero, then the server unicasts DHCPOFFER and DHCPACK
+		 * messages to the address in 'ciaddr'.
+		 */
+		dst_addr.sin_addr = ciaddr;
+	} else if (ntohs(msg->flags) & DHCPV4_MSG_BROADCAST) {
+		/* If 'giaddr' is zero and 'ciaddr' is zero, and the broadcast
+		 * bit is set, then the server broadcasts DHCPOFFER and DHCPACK
+		 * messages to 0xffffffff.
+		 */
+		dst_addr.sin_addr = *net_ipv4_broadcast_address();
+	} else if (yiaddr != NULL) {
+		/* If the broadcast bit is not set and 'giaddr' is zero and
+		 * 'ciaddr' is zero, then the server unicasts DHCPOFFER and
+		 * DHCPACK messages to the client's hardware address and 'yiaddr'
+		 * address.
+		 */
+		struct net_eth_addr hwaddr;
+
+		memcpy(&hwaddr, msg->chaddr, sizeof(hwaddr));
+		net_arp_update(ctx->iface, yiaddr, &hwaddr, false, true);
+		dst_addr.sin_addr = *yiaddr;
+	} else {
+		NET_ERR("Unspecified destination address.");
+		return -EDESTADDRREQ;
+	}
+
+	ret = zsock_sendto(ctx->sock, reply, len, 0, (struct sockaddr *)&dst_addr,
+			   sizeof(dst_addr));
+	if (ret < 0) {
+		return -errno;
+	}
+
+	return 0;
+}
+
+static int dhcpv4_send_offer(struct dhcpv4_server_ctx *ctx, struct dhcp_msg *msg,
+			     struct in_addr *addr, uint32_t lease_time,
+			     struct dhcpv4_parameter_request_list *params)
+{
+	uint8_t reply[NET_IPV4_MTU];
```

Comment:
Could we do with smaller buffer, it is now 576 bytes long?
One option could be define the buffer size in kconfig and let user to adjust it.

Reply:
I didn't see any specific requirement for the server in the RFC2131 (it's quite chaotic to be honest), but for the client it says:
> The 'options' field is now variable length. A DHCP client must be
   prepared to receive DHCP messages with an 'options' field of at least
   length 312 octets.  This requirement implies that a DHCP client **must
   be prepared to receive a message of up to 576 octets** , the minimum IP
   datagram size an IP host must be prepared to accept
   
Hence my rationale to use IPv4_MTU for the buffer size. In practice, I see in wireshark that Discover/Request sent by th client are 342 bytes long, Offer/Ack sent in response are 304 bytes, so there is a little room for improvement if we find it valuable to optimize a bit. WDYT, is it worth it?

Reply:
Dunno, and this can be optimized later too. Perhaps we can leave that like this and change it later if needed.